api-registration/server/server.js

```diff
@@ -0,0 +1,66 @@
+/* eslint-disable no-unused-vars -- REMOVE ME */
+import 'dotenv/config';
+import pg from 'pg';
+import argon2 from 'argon2';
+import express from 'express';
+import { ClientError, errorMiddleware } from './lib/index.js';
+
+const db = new pg.Pool({
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+app.use(express.json());
+
+async function checkUsername(username, res) {
+  const sql = `
+      SELECT "username" FROM "users"
+      WHERE "username" = $1;`;
+  const params = [username];
+  const result = await db.query(sql, params);
+  const user = result.rows[0];
+  if (user) {
+    throw new ClientError(400, `Username '${username}' already exists. Please try another username.`);
+  }
+}
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+    await checkUsername(username, res);
+    /* TODO:
+     * Hash the user's password with `argon2.hash()`
+     * Insert the user's "username" and "hashedPassword" into the "users" table.
+     * Respond to the client with a 201 status code and the new user's "userId", "username", and "createdAt" timestamp.
+     * Catch any errors.
+     *
+     * Hint: Insert statements can include a `returning` clause to retrieve the insterted row(s).
+     */
+    const sql = `
+      INSERT INTO "users" ("username", "hashedPassword")
+      VALUES ($1, $2)
+      RETURNING *;`;
+    const hashedPassword = await argon2.hash(password);
+    const params = [username, hashedPassword];
+    const result = await db.query(sql, params);
+    const user = result.rows[0];
+    if (user) {
```

Comment:
You don't give the user any feedback if user is undefined. However, it should never be undefined since if there is an error inserting, an exception will be thrown.